### PR TITLE
feat!: Use Template Classes and Google Test

### PR DESCRIPTION
**Pull request by HazyFish**

### CODE REVIEW

File: tests/LinkedListTest.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#include <gtest/gtest.h>
+#include "LinkedList.cpp"
+
+TEST(LinkedListTest, InstantiateEmptyList) {
```
Comment (EYH0602) [inline tests/LinkedListTest.cpp:4]: are we exposing 3 tests for students?

File: lib/Stack.hpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+class Stack
+{
+private:
+    LinkedList<T> _llist;
+
+public:
+    Stack();
+    Stack(LinkedList<T> &&llist);
+    size_t size();
+    std::optional<T> top();
```
Comment (EYH0602) [inline lib/Stack.hpp:18]: I'll add a slide on optional (or sum type in general) on next week discussion

**Review by EYH0602 — APPROVED** [review]

File: tests/LinkedListTest.cpp (hunk 1 above)
Comment (HazyFish) [reply tests/LinkedListTest.cpp:4]: I'm thinking of showing test results from the 3 test C++ files the students submit (regardless of what they add/modify), but they will have 0 weight.

The final score will be graded on the test cases written by us only. (We don't use their `tests/*.cpp` when running our grader.)

What do you think?

Comment (EYH0602) [reply tests/LinkedListTest.cpp:4]: totally agree

**Pull request merged**